proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,532 @@
+
+One place where WebAssembly and JavaScript invariants don't naturally
+line up is *reference types*.
+
+In particular, the JS Typed Object proposal `ref` Value Type objects are
+[defined to preserve the usual "nominal typing" of JavaScript](https://github.com/tschneidereit/proposal-typed-objects/blob/master/explainer.md#type-checking-for-struct-type-references).
+In particular, field type implies a specific prototype *identity*. In contrast,
+the wasm GC proposal's [static subtyping relation](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#subtyping)
+is *structural*, ensuring only field types and layouts.
+
+To see this problem, consider the following JS code which hands a mutable `Rect`
+typed object to WebAssembly and then inspects it afterwards.
+
+```js
+const Point = new StructType([{name: "x", type: int32}, {name: "y", type: int32}]);
+const Rect = new StructType([{name: "p1", type: Point.ref}, {name: "p2", type: Point.ref}]);
+var r = new Rect(new Point(1,1), new Point(2,2));
+
+wasmInstance.exports.goWild(r);
+
+assert(r.p1 instanceof Point);
+assert(r.p2 instanceof Point);
+```
+
+Without knowing anything about `wasmInstance`, can JS assume *based solely on
+the field types of `Rect`* that the asserts hold?
+
+If the JS `ref` Value Types were unified with the WebAssembly
+[`ref` type constructor](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types),
+the answer would be "no": `wasmInstance` could assign any typed object with two
+`i32` fields, including those with a `null` `__proto__` and thus no `x` and `y`
+property names. This would be surprising from a JS perspective and would also
+give up future WebAssembly Host Binding optimizations that are mentioned
+at the end of this section.
+
+To avoid problem, as a starting point, the following WebAssembly module:
+
+```wat
+;; example3.wat --> example3.wasm
+(module
+    (type $P (struct (field i32) (field i32)))
+    (type $R (struct (field $x1 (ref $P)) (field $x2 (ref $P))))
+    (func (export "goWild") (param (ref $R))
+        (struct.set $R $x1
+            (get_local 0)
+            (struct.new $P (i32.const 10) (i32.const 20)))
+    )
+)
+```
+
+will throw when passed a `Point` typed object because of the attempt to
+reinterpret the *nominal* field types of `Rect` with the *structural*
+field types of `$R`.
+
+```js
+WebAssembly.instantiateStreaming(fetch('example3.wasm'))
+.then(({instance}) => {
+    instance.exports.goWild(new Rect(new Point(1,2), new Point(3,4)));  // throws
+});
+```
+
+For the call to succeed, the WebAssembly module must intead *import* the
+exact Type Definition object and use that type import in all relevant
+function signatures:
+
+```wat
+;; example4.wat --> example4.wasm
+(module
+    (type $Point (import "" "Point") (eq (struct (field $x i32) (field $y i32))))
+    (type $Rect (struct (field (ref $Point)) (field (ref $Point))))
+    (func (export "goWild") (param (ref $Rect))
+        (struct.set $R $x1
```

Comment:
Do you mean $Rect and $x here?

Reply:
You're right I did mean $Rect instead of $R, and but I did mean $x1, I just forgot to add this field name to $Rect's first field :)